File: map-gen-script/src/map_generator/topologies/star_shape.py
import random
from .base_topology import BaseTopology
from typing import Iterator
import copy
from src.map_generator.models.path_info import PathInfo, Coord
from src.utils.geometry import add_vectors, FORWARD_X, FORWARD_Z, BACKWARD_X, BACKWARD_Z

class StarShapeTopology(BaseTopology):
    """
    Tạo ra một đường đi liên tục theo đường viền của một ngôi sao 5 cánh (pixel art).
    Đường đi được thiết kế để có Eulerian Path, cho phép robot di chuyển hết toàn bộ map.
    """
    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [REWRITTEN] Tạo biến thể bằng cách tăng dần kích thước từ params gốc.
        """
        base_size = int(params.get('star_size', 2))

        for i in range(max_variants):
            variant_params = copy.deepcopy(params)
            # Tăng dần kích thước cho mỗi biến thể
            variant_params['star_size'] = base_size + i
            # Cập nhật cả path_length để Placer có yêu cầu tương ứng
            if 'path_length' in variant_params:
                variant_params['path_length'] = int(params.get('path_length', 3)) + i

            yield self.generate_path_info(grid_size, variant_params)

    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        print("    LOG: Generating 'star_shape' (outline) topology...")
        size = params.get('star_size', 3)  # Kích thước mỗi đoạn của ngôi sao
        # [NÂNG CẤP] Đọc path_mode để quyết định loại đường đi trả về.
        # 'full_outline' (mặc định) cho các Placer thông thường.
        # 'straight_segment' cho PathSearchingPlacer.
        path_mode = params.get('path_mode', 'full_outline')
        if size < 2: size = 2

        center_x = grid_size[0] // 2
        center_z = grid_size[2] // 2

        # --- PHẦN 1: TẠO HÌNH DẠNG (placement_coords) ---
        current_pos = (center_x, 0, center_z - size * 2)
        path_coords = [current_pos]

        # Ánh xạ hướng chuỗi sang vector
        direction_map = {
            'right': FORWARD_X,
            'down': FORWARD_Z,
            'left': BACKWARD_X,
            'up': BACKWARD_Z,
        }

        # [MỚI] Lưu trữ các đoạn thẳng để chọn làm đường đi chính
        straight_segments = []
        # Danh sách các đoạn di chuyển (hướng, số bước) để vẽ viền ngôi sao
        segments = [
            ('right', size),     # Đi đến đỉnh trên-phải
            ('down', size),      # Đi xuống cánh phải
            ('left', size),      # Đi vào trong
            ('down', size),      # Đi xuống đỉnh dưới-phải
            ('left', size),      # Đi qua đỉnh dưới
            ('up', size),        # Đi lên đỉnh dưới-trái
            ('left', size),      # Đi ra cánh trái
            ('up', size),        # Đi lên đỉnh trên-trái
            ('right', size),     # Đi vào trong
            ('up', size - 1),    # Đi lên gần đỉnh trên để nối vòng lặp
        ]

        for direction_key, steps in segments:
            segment_start_pos = current_pos
            move_vector = direction_map[direction_key]
            current_segment = [segment_start_pos]
            for _ in range(steps):
                current_pos = add_vectors(current_pos, move_vector)
                path_coords.append(current_pos)
                current_segment.append(current_pos)
            straight_segments.append(current_segment)

        # Loại bỏ các điểm trùng lặp nhưng vẫn giữ nguyên thứ tự
        placement_coords = list(dict.fromkeys(path_coords))

        # --- PHẦN 2: [NÂNG CẤP] TẠO ĐƯỜNG ĐI (path_coords) DỰA TRÊN path_mode ---
        if path_mode == 'straight_segment':
            # Chế độ cũ: Trích xuất một đoạn thẳng cho PathSearchingPlacer.
            # Chọn một trong các đoạn thẳng dài nhất (các "cánh" của ngôi sao).
            if straight_segments:
                # Lọc ra các đoạn có độ dài `size+1` (các cánh ngoài)
                long_segments = [seg for seg in straight_segments if len(seg) == size + 1]
                final_path = random.choice(long_segments) if long_segments else random.choice(straight_segments)
            else:
                final_path = placement_coords # Fallback
        else:
            # Chế độ mới (mặc định): Sử dụng toàn bộ đường viền làm đường đi.
            final_path = placement_coords

        # [NEW] Tạo metadata cho Placers
        metadata = {
            "segments": straight_segments,
            "star_size": size,
        }

        return PathInfo(
            start_pos=final_path[0],
            target_pos=final_path[-1],
            path_coords=final_path,
            # `placement_coords` luôn là toàn bộ cấu trúc để render nền.
            placement_coords=placement_coords,
            metadata=metadata
        )

File: apps/map-builder-app/src/map-generator/topologies/StarShape.ts

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';
import { addVectors, FORWARD_X, FORWARD_Z, BACKWARD_X, BACKWARD_Z } from '../utils/geometry';

export class StarShapeTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    const baseSize = params.star_size || 2;

    for (let i = 0; i < maxVariants; i++) {
        const variantParams = { ...params };
        variantParams.star_size = baseSize + i;
        if (params.path_length) {
             variantParams.path_length = (params.path_length || 3) + i;
        }
        yield this.generatePathInfo(gridSize, variantParams);
    }
  }

  generatePathInfo(
    gridSize: [number, number, number],
    params: Record<string, any>
  ): IPathInfo {
    let size = params.star_size || 3;
    const pathMode = params.path_mode || 'full_outline';
    if (size < 2) size = 2;

    const centerX = Math.floor(gridSize[0] / 2);
    const centerZ = Math.floor(gridSize[2] / 2);

    let currentPos: Coord = [centerX, 0, centerZ - size * 2];
    const pathCoords: Coord[] = [currentPos];

    // Map strings to vectors
    const directionMap: Record<string, Coord> = {
        'right': FORWARD_X,
        'down': FORWARD_Z,
        'left': BACKWARD_X,
        'up': BACKWARD_Z,
    };

    const straightSegments: Coord[][] = [];
    
    // Star outline logic
    const segmentsDef = [
        {dir: 'right', steps: size},
        {dir: 'down', steps: size},
        {dir: 'left', steps: size},
        {dir: 'down', steps: size}, // right-bottom corner
        {dir: 'left', steps: size}, // bottom peak
        {dir: 'up', steps: size},   // left-bottom corner
        {dir: 'left', steps: size}, // left arm
        {dir: 'up', steps: size},   // left-top corner
        {dir: 'right', steps: size}, // top-left inward
        {dir: 'up', steps: size - 1}, // connect back
    ];

    for (const def of segmentsDef) {
        const moveVector = directionMap[def.dir];
        const currentSegment = [currentPos];
        for (let i = 0; i < def.steps; i++) {
            currentPos = addVectors(currentPos, moveVector);
            pathCoords.push(currentPos);
            currentSegment.push(currentPos);
        }
        straightSegments.push(currentSegment);
    }

    // Deduplicate
    const uniqueCoordsSet = new Set<string>();
    const placementCoords: Coord[] = [];
    for(const c of pathCoords) {
        const key = c.join(',');
        if(!uniqueCoordsSet.has(key)) {
            uniqueCoordsSet.add(key);
            placementCoords.push(c);
        }
    }

    // Path selection logic
    let finalPath: Coord[];
    if (pathMode === 'straight_segment') {
        const longSegments = straightSegments.filter(s => s.length === size + 1);
        if (longSegments.length > 0) {
            finalPath = longSegments[Math.floor(Math.random() * longSegments.length)];
        } else if (straightSegments.length > 0) {
            finalPath = straightSegments[Math.floor(Math.random() * straightSegments.length)];
        } else {
             finalPath = placementCoords;
        }
    } else {
        finalPath = placementCoords;
    }

    // Metadata
    const corners: Coord[] = [];
    for(const seg of straightSegments) {
        if(seg.length > 0) corners.push(seg[0]);
    }

    const branches: Coord[][] = [];
    for(let i=0; i<straightSegments.length - 1; i+=2) {
        // combine seg[i] and seg[i+1] skipping join point
        const arm = [...straightSegments[i], ...straightSegments[i+1].slice(1)];
        branches.push(arm);
    }

    const semanticPositions: Record<string, any> = {
        center: [centerX, 0, centerZ],
        start_point: finalPath[0],
        end_point: finalPath[finalPath.length - 1],
    };
    
    // Arm endpoints
    branches.forEach((branch, i) => {
        if (branch.length > 0) {
            const tipIdx = Math.floor(branch.length / 2);
            semanticPositions[`arm_${i}_end`] = branch[Math.min(tipIdx, branch.length - 1)];
        }
    });

    semanticPositions.optimal_start = 'center';
    semanticPositions.optimal_end = 'arm_0_end';
    
    const metadata = {
        topology_type: 'star_shape',
        segments: straightSegments,
        branches: branches,
        corners: corners,
        star_size: size,
        semantic_positions: semanticPositions
    };

    return {
        start_pos: finalPath[0],
        target_pos: finalPath[finalPath.length - 1],
        path_coords: finalPath,
        placement_coords: placementCoords,
        obstacles: [],
        metadata: metadata
    };
  }
}
